compiler/rustc_target/src/spec/tests/tests_impl.rs

```diff
@@ -109,20 +112,63 @@ impl Target {
             );
         }
 
-        assert!(
-            (self.pre_link_objects_self_contained.is_empty()
-                && self.post_link_objects_self_contained.is_empty())
-                || self.link_self_contained != LinkSelfContainedDefault::False
-        );
+        if self.link_self_contained == LinkSelfContainedDefault::False {
+            assert!(
+                self.pre_link_objects_self_contained.is_empty()
+                    && self.post_link_objects_self_contained.is_empty()
+            );
+        }
 
         // If your target really needs to deviate from the rules below,
         // except it and document the reasons.
         // Keep the default "unknown" vendor instead.
         assert_ne!(self.vendor, "");
+        assert_ne!(self.os, "");
         if !self.can_use_os_unknown() {
             // Keep the default "none" for bare metal targets instead.
             assert_ne!(self.os, "unknown");
         }
+
+        // Check dynamic linking stuff
+        // FIXME: How does bpf do dynamic linking?
+        if self.os == "none" && self.arch != "bpf" {
+            assert!(!self.dynamic_linking);
```

Comment:
cc @alessandrod 
How does dynamic linking works with BPF? Who loads the libraries?

Reply:
When targeting the kernel vm, nothing does. When targeting user space vms (like rbpf), those can load libraries, or use unresolved symbols to implement "syscalls".